bxserum/connections/ws.go

```diff
@@ -5,84 +5,325 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/bloXroute-Labs/serum-client-go/utils"
 	"github.com/gorilla/websocket"
 	"github.com/sourcegraph/jsonrpc2"
 	"google.golang.org/protobuf/encoding/protojson"
 	"google.golang.org/protobuf/proto"
-	"strings"
+	"sync"
+	"time"
 )
 
-func WSRequest[T proto.Message](conn *websocket.Conn, request []byte, response T) error {
-	err := sendWS(conn, request)
+const (
+	handshakeTimeout       = 5 * time.Second
+	subscriptionBuffer     = 1000
+	unsubscribeGracePeriod = 3 * time.Second
+)
+
+type WS struct {
+	messageM      sync.Mutex
+	subscriptionM sync.RWMutex
+	requestID     *utils.RequestID
+	conn          *websocket.Conn
+	ctx           context.Context
+	cancel        context.CancelFunc
+	err           error
+	writeCh       chan []byte
+
+	requestMap      map[uint64]requestTracker
+	subscriptionMap map[string]subscriptionEntry
+}
+
+func NewWS(endpoint string) (*WS, error) {
+	dialer := websocket.Dialer{HandshakeTimeout: handshakeTimeout}
+	conn, _, err := dialer.Dial(endpoint, nil)
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ws := &WS{
+		requestID:       utils.NewRequestID(),
+		conn:            conn,
+		ctx:             ctx,
+		cancel:          cancel,
+		writeCh:         make(chan []byte, 100),
+		requestMap:      make(map[uint64]requestTracker),
+		subscriptionMap: make(map[string]subscriptionEntry),
+	}
+	go ws.readLoop()
+	go ws.writeLoop()
+	return ws, nil
+}
+
+func (w *WS) readLoop() {
+	defer w.cancel()
+
+	for {
+		// if lock is held by another processing routine, then wait for release before processing any further messages
+		// typically, messages can be dispatched independently, but in some cases it may be important to finish processing a single message before allowing the socket to read any more (e.g. subscription response before processing potential updates)
+		w.messageM.Lock()
+		w.messageM.Unlock()
+
+		_, msg, err := w.conn.ReadMessage()
+		if err != nil {
+			_ = w.Close(err)
+			return
+		}
+
+		// try response format first
+		var response jsonrpc2.Response
+		err = json.Unmarshal(msg, &response)
+		if err == nil && (response.Result != nil || response.Error != nil) {
+			w.processRPCResponse(response)
+			continue
+		}
+
+		// if not, try subscription format
+		var update jsonrpc2.Request
```

Comment:
jsonrpc2.Request ? should it be FeedUpdate ?

Reply:
yeah, so "update" and "request" are the same shape as each other in JSON-RPC. There are only two formats in the JSON-RPC spec, "request" and "response. "response" is only ever in response to a "request", so subscription updates can't be of the "response" format. Instead, it's a "request" without a `requestID`, which indicates that it doesn't expect a "response" in return.

Example:

**Request:**
```
wscat -c ws://127.0.0.1:9000/ws --wait 1000 --execute '{"jsonrpc": "2.0", "id": 1, "method": "subscribe", "params": ["GetOrderbooksStream", {"market": "SOL/USDC", "limit": 10}]}'
```

**Response** (subscribe confirmation):
```
{"id":1,"result":"3c911616-0a8c-4184-a733-138c163f5c03","jsonrpc":"2.0"}
```

**Event** (update "request" with no ID):
```
{"method":"subscribe","params":{"subscription":"3c911616-0a8c-4184-a733-138c163f5c03","result":{"blockHeight":"141322189","orderbook":{"market":"SOL/USDC","marketAddress":"9wFFyRfZBsuAha4YcuxcXLKwMxJR43S7fPfQLusDBzvT","bids":[{"price":32.916,"size":1289.4},{"price":32.93,"size":3.3},{"price":32.943,"size":1507},{"price":32.956,"size":643.9},{"price":32.976,"size":321.7},{"price":32.986,"size":160.8},{"price":33.091,"size":603.2},{"price":33.135,"size":200},{"price":33.137,"size":90.4},{"price":33.138,"size":11.1}],"asks":[{"price":33.165,"size":160.8},{"price":33.175,"size":321.6},{"price":33.195,"size":643.2},{"price":33.2,"size":233.1},{"price":33.217,"size":0.1},{"price":33.218,"size":3},{"price":33.227,"size":0.1},{"price":33.23,"size":90.4},{"price":33.235,"size":1286.5},{"price":33.25,"size":1615.6}]}}},"jsonrpc":"2.0"}
```